pkg/controller/certificates/trigger/policies/policies_test.go

```diff
@@ -506,3 +511,265 @@ func TestDefaultPolicyChain(t *testing.T) {
+			Encoding:       cmapi.PKCS8,
+		},
+		EncodeUsagesInRequest: pointer.Bool(true),
+		RevisionHistoryLimit:  pointer.Int32(2),
+	}}
+	originalSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{Name: "something",
+			Annotations: map[string]string{
+				cmapi.IssuerNameAnnotationKey:  "testissuer",
+				cmapi.IssuerKindAnnotationKey:  "IssuerKind",
+				cmapi.IssuerGroupAnnotationKey: "group.example.com",
+			},
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: staticFixedPrivateKey,
+			corev1.TLSCertKey:       internaltest.MustCreateCert(t, staticFixedPrivateKey, originalCert),
+		},
+	}
+	originalCR := &cmapi.CertificateRequest{Spec: cmapi.CertificateRequestSpec{
+		IssuerRef: cmmeta.ObjectReference{
+			Name:  "testissuer",
+			Kind:  "IssuerKind",
+			Group: "group.example.com",
+		},
+		Request: internaltest.MustGenerateCSRImpl(t, staticFixedPrivateKey, originalCert),
+		IsCA:    true,
+		Usages: []cmapi.KeyUsage{
+			cmapi.UsageSigning,
+			cmapi.UsageDigitalSignature,
+			cmapi.UsageKeyEncipherment,
+		},
+		Duration: &metav1.Duration{Duration: 30 * 24 * time.Hour},
+	}}
+
+	tests := []struct {
+		noCR  bool // When true, pretend that the CertificateRequest was not found.
+		noSec bool // When true, pretend that the Secret was not found.
+
+		// Note that we do not test with changes to the CertificateRequest
+		// because this resource is supposed to be immutable.
+		cert   func(c *Cert)          // Nil if no change is to be made.
+		secret func(c *corev1.Secret) // Nil if no change is to be made.
+
+		reissue bool
+	}{
+		// Happy case: with no change to the Certificate nor the Secret, and
+		// with or without the CertificateRequest available, no re-issuance is
+		// expected.
+		{noCR: false, reissue: false},
+		{noCR: true, reissue: false},
+
+		// When the Secret is missing or broken, a re-issuance is expected.
+		{noCR: true, noSec: true, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data = nil }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data[corev1.TLSPrivateKeyKey] = nil }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data[corev1.TLSCertKey] = nil }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data[corev1.TLSPrivateKeyKey] = []byte("invalid") }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data[corev1.TLSCertKey] = []byte("invalid") }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Data[corev1.TLSPrivateKeyKey] = internaltest.MustCreatePEMPrivateKey(t) }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Annotations["cert-manager.io/issuer-name"] = "change" }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Annotations["cert-manager.io/issuer-kind"] = "change" }, reissue: true},
+		{noCR: true, secret: func(s *corev1.Secret) { s.Annotations["cert-manager.io/issuer-group"] = "change" }, reissue: true},
```

Comment:
@irbekrm I reproduced the test cases from `TestDefaultPolicyChain` here. I like the compact form of these test cases but I realize that it is missing the check for the reason/message... I guess I'll have to add one more line to each test case, e.g.:

```go
{noCR: true, secret: func(s *corev1.Secret) { s.Annotations["cert-manager.io/issuer-group"] = "change" }, reissue: true
	message: "IncorrectIssuer: Issuing certificate as Secret was previously issued by IssuerKind.new.example.com/testissuer"},
```

I'll do that tomorrow morning.

Reply:
Done in 9409f8a0ba216d33ceb119dc97baca8e35257d9a.